python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,480 @@
+
+        # if not dtype is specified, we try to get it from the model
+        if dtype is None:
+            self.dtype = get_dtype_from_model_func(func=model,
+                                                   default=np.float32)
+        else:
+            self.dtype = np.dtype(dtype)
+
+        self.background, self.N, self.M, _ = \
+            input_to_cuml_array(data, order=self.order,
+                                convert_to_dtype=self.dtype)
+
+        self.nsamples = 2 * self.M + 2 ** 11 if nsamples is None else nsamples
+
+        self.max_samples = 2 ** 30
+
+        # restricting maximum number of samples for memory and performance
+        # value being checked, right now based on mainline SHAP package
+        self.max_samples = 2 ** 30
+        if self.M <= 30:
+            self.max_samples = 2 ** self.M - 2
+            if self.nsamples > self.max_samples:
+                self.nsamples = self.max_samples
+
+        if isinstance(data, pd_df) or isinstance(data, cu_df):
+            self.feature_names = data.columns.to_list()
+        else:
+            self.feature_names = [None for _ in range(len(data))]
+
+        # seeing how many exact samples from the powerset we can enumerate
+        # todo: optimization for larger sizes by generating diagonal
+        # and gpu lexicographical-binary numbers generation
+        cur_nsamples = self.M
+        r = 1
+        while cur_nsamples < self.nsamples:
+            if has_scipy():
+                from scipy.special import binom
+                cur_nsamples += int(binom(self.M, r))
+            else:
+                cur_nsamples += int(binomCoef(self.M, r))
+            r += 1
+
+        # see if we need to have randomly sampled entries in our mask
+        # and combinations matrices
+        self.nsamples_random = max(self.nsamples - cur_nsamples, 0)
+
+        # using numpy powerset and calculations for initial version
+        # cost is incurred only once, and generally we only generate
+        # very few samples if M is big.
+        mat, weight = powerset(self.M, r, self.nsamples, dtype=self.dtype)
+        weight /= np.sum(weight)
+
+        self.mask, *_ = input_to_cuml_array(mat)
+        self.nsamples_exact = len(self.mask)
+
+        self.weights = cp.empty(self.nsamples, dtype=self.dtype)
+        self.weights[:self.nsamples_exact] = cp.array(weight)
+
+        self.synth_data = None
+
+        self.expected_value = self.link_fn(cp.mean(model(self.background)))
+
+        self.random_state = random_state
+
+    def explain(self,
+                X,
+                nsamples=None,
+                l1_reg='auto'):
+        shap_values = cp.zeros((1, self.M), dtype=self.dtype)
+
+        # allocating combinations array once for multiple explanations
+        if self.synth_data is None:
+            self.synth_data = CumlArray.zeros(
+                shape=(self.N * self.nsamples, self.M),
+                dtype=np.float32,
```

Comment:
I see some support for doubles, but not this seems to be all float32? Maybe it gets overwritten in that case?